### PR TITLE
ABI encoder fixes and test.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/codegen/ABIFunctions.cpp [hunk 1]
```diff
@@ -1190,10 +1190,7 @@ size_t ABIFunctions::headSize(TypePointers const& _targetTypes)
 		if (t->isDynamicallyEncoded())
 			headSize += 0x20;
 		else
-		{
-			solAssert(t->calldataEncodedSize() > 0, "");
```
Comment (chriseth) [inline libsolidity/codegen/ABIFunctions.cpp:1194]: It turns out, calldata encode size can actually be zero e.g. for an empty struct.

File: libsolidity/codegen/ABIFunctions.cpp [hunk 2]
```diff
@@ -538,7 +538,7 @@ string ABIFunctions::abiEncodingFunctionSimpleArray(
 						mstore(pos, sub(tail, headStart))
 						tail := <encodeToMemoryFun>(<arrayElementAccess>, tail)
 						srcPtr := <nextArrayElement>(srcPtr)
-						pos := add(pos, <elementEncodedSize>)
```
Comment (chriseth) [inline libsolidity/codegen/ABIFunctions.cpp:541]: This failed here because we only encode a pointer and not the actual data.

**Comment by chriseth** [comment]
Added tests for empty structs.

File: libsolidity/codegen/ABIFunctions.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/codegen/ABIFunctions.cpp:541]: `dynamicBase` says whether it is a fixed-length or dynamic-length?

File: libsolidity/codegen/ABIFunctions.cpp [hunk 3]
```diff
@@ -832,7 +832,7 @@ string ABIFunctions::abiEncodingFunctionStruct(
 			}
 			memberTempl("encodingOffset", toCompactHexWithPrefix(encodingOffset));
 			encodingOffset += dynamicMember ? 0x20 : memberTypeTo->calldataEncodedSize();
-			memberTempl("abiEncode", abiEncodingFunction(*memberTypeFrom, *memberTypeTo, _encodeAsLibraryTypes, false));
+			memberTempl("abiEncode", abiEncodingFunction(*memberTypeFrom, *memberTypeTo, _encodeAsLibraryTypes, true));
```
Comment (axic) [inline libsolidity/codegen/ABIFunctions.cpp]: The `true` here seems to refer to the `_compacted` parameter which only seems to be used for function type encoding.  Which then seems to mean whether the function type is a single stack item or two items (addr, selector), where true means single.

Is that correct? Takes quite a lot to trace it and hence understanding what is going on.

File: test/libsolidity/ABIEncoderTests.cpp [hunk 4]
```diff
@@ -462,6 +462,87 @@ BOOST_AUTO_TEST_CASE(structs)
...
+			function f() returns (uint, S, uint) {
+				e(7, s, 8);
+				return (7, s, 8);
+			}
+		}
+	)";
+
+	NEW_ENCODER(
+		compileAndRun(sourceCode, 0, "C");
+		bytes encoded = encodeArgs(7, 8);
```
Comment (axic) [inline test/libsolidity/ABIEncoderTests.cpp:481]: Hm, is the struct encoded in place?

Comment (chriseth) [reply test/libsolidity/ABIEncoderTests.cpp:481]: yes, because it is statically sized

File: libsolidity/codegen/ABIFunctions.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp:541]: whether it is encoded statically or dynamically

File: libsolidity/codegen/ABIFunctions.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp]: Yes, that's true. Unfortunately, we have to pass this parameter through all calls. Do you have a better idea?

Comment (axic) [reply libsolidity/codegen/ABIFunctions.cpp]: Why do we have two versions of the stack for this type?

Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp]: If the type is stored on the stack, it always uses two slots, but if it is stored in memory, those are combined. It makes sense to not split and recombine it in the ABI encoding if it is just copied somwhere, for example if you want to encode a memory array of function type.

Comment (axic) [reply libsolidity/codegen/ABIFunctions.cpp]: So perhaps a better name instead of `compact` is `fromMemory`/`fromStack` ?

Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp]: Yes, I think I did something like that in the decoder.

**Comment by chriseth** [comment]
Can this be merged, @axic?

File: libsolidity/codegen/ABIFunctions.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/codegen/ABIFunctions.cpp]: Can you rename it? If not int his PR in another one before we forget.

Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp]: Renamed.

**Review by axic — APPROVED** [review]

**Pull request merged**